bootstrap/api/v1alpha1/rke2config_types.go

```diff
@@ -18,47 +18,624 @@ package v1alpha1
+// - Used in the RKE2Config CRD
+// - Used to generate a registries.yaml file for the RKE2 command line
+// In the CRD, CASecret attribute is a reference to a Secret in the Kube API, while in the registries.yaml CASecret is marshalled into ca_file which a file path on the system
+// This means the Registry object will need to be copied and processed before marshalling
+
+// TLSConfig contains the CA/Cert/Key used for a registry
+type TLSConfig struct {
+
+	// CASecret is a reference to a secret containing the Certificate Authority used to verify the registry's certificate
+	//+optional
+	CASecret SecretRef `json:"caSecret,omitempty"`
+
+	// CertSecret is a reference to a secret containing the Certificate used to authenticate to the registry
+	//+optional
+	CertSecret SecretRef `json:"certSecret,omitempty"`
+
+	// KeySecret is a reference to a secret containing the Key for the certificate used to authenticate to the registry
+	//+optional
+	KeySecret SecretRef `json:"keySecret,omitempty"`
+
+	//InsecureSkipVerify may be set to true to skip verifying the registry's certificate
+	//+optional
+	InsecureSkipVerify bool `yaml:"insecure_skip_verify" json:"insecureSkipVerify,omitempty"`
+}
+
+type EtcdS3 struct {
+
+	// Endpoint S3 endpoint url (default: "s3.amazonaws.com")
+	// +required
+	Endpoint string `json:"endpoint"`
+
+	// EndpointCA references the Secret that contains a custom CA that should be trusted to connect to S3 endpoint
+	// +optional
+	EndpointCA SecretRef `json:"endpointCA,omitempty"`
+
+	// SkipSslVerify Disables S3 SSL certificate validation
+	// +optional
+	SkipSslVerify string `json:"skipSslVerify,omitempty"`
+
+	// S3CredentialSecret is a reference to a Secret containing the Access Key and Secret Key necessary to access the target S3 Bucket
+	//+required
+	S3CredentialSecret SecretRef `json:"S3CredentialSecret"`
+
+	// Bucket S3 bucket name
+	// +optional
+	Bucket string `json:"bucket,omitempty"`
+
+	// Region S3 region / bucket location (optional) (default: "us-east-1")
+	// +optional
+	Region string `json:"region,omitempty"`
+
+	// Folder S3 folder
+	// +optional
+	Folder string `json:"folder,omitempty"`
+}
+
+type ComponentConfig struct {
+
+	// ExtraEnv is a map of environment variables to pass on to a Kubernetes Component command
+	//+optional
+	ExtraEnv map[string]string `json:"extraEnv,omitempty"`
+
+	// ExtraArgs is a map of command line arguments to pass to a Kubernetes Component command
+	//+optional
+	ExtraArgs map[string]string `json:"extraArgs,omitempty"`
+
+	//ExtraMounts is a map of volume mounts to be added for the Kubernetes component Pod
+	//+optional
+	ExtraMounts map[string]string `json:"extraMounts,omitempty"`
+
+	//OverrideImage is a string that references a container image to override the default one for the Kubernetes Component
+	//+optional
+	OverrideImage string `json:"overrideImage,omitempty"`
+}
+
+// SecretRef references a secret using a namespace and a name
+type SecretRef struct {
```

Comment:
this structure can be replaced with objectrefence structure https://pkg.go.dev/k8s.io/api/core/v1#ObjectReference